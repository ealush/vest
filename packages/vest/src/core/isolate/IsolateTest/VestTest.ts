import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
