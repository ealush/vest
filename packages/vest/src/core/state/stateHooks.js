import ctx from 'ctx';

const getStateRef = () => ctx.use().stateRef;

export function usePending() {
  return getStateRef().pending();
}
export function useSuiteId() {
  return getStateRef().suiteId();
}
export function useTestCallbacks() {
  return getStateRef().testCallbacks();
}
export function useTestObjects() {
  return getStateRef().testObjects();
}
export function useSkippedTests() {
  return getStateRef().skippedTests();
}
export function useOptionalFields() {
  return getStateRef().optionalFields();
}
