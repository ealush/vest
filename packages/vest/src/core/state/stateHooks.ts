import ctx from 'ctx';

export function useCarryOverTests() {
  return useStateRef().carryOverTests();
}
export function usePending() {
  return useStateRef().pending();
}
export function useLagging() {
  return useStateRef().lagging();
}
export function useSuiteId() {
  return useStateRef().suiteId();
}
export function useTestCallbacks() {
  return useStateRef().testCallbacks();
}
export function useTestObjects() {
  return useStateRef().testObjects();
}
export function useSkippedTests() {
  return useStateRef().skippedTests();
}
export function useOptionalFields() {
  return useStateRef().optionalFields();
}
export function useStateRef() {
  return ctx.use()!.stateRef!;
}
