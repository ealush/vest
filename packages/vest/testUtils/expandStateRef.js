export default function expandStateRef(stateRef) {
  const state = {};
  for (const key in stateRef) {
    const [value] = stateRef[key]();
    state[key] = value;
  }
  return state;
}
