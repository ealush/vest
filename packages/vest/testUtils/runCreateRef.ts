import createState from 'vast';

import createStateRef from 'createStateRef';

export default (
  state?: ReturnType<typeof createState>
): ReturnType<typeof createStateRef> =>
  createStateRef(state ? state : createState(), {
    suiteId: '1000',
  });
