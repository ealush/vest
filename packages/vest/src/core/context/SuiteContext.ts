import { createCascade } from 'context';
import { assign, TinyState, tinyState, cache, CacheApi } from 'vest-utils';
import { DynamicValue } from 'vest-utils/src/utilityTypes';

import { IsolateTest } from 'IsolateTest';
import { Modes } from 'Modes';

export const SuiteContext = createCascade<CTXType>((ctxRef, parentContext) => {
  if (parentContext) {
    return null;
  }

  return assign(
    {
      inclusion: {},
      mode: tinyState.createTinyState<Modes>(Modes.EAGER),
      suiteParams: [],
      testMemoCache,
    },
    ctxRef
  );
});

type CTXType = {
  inclusion: Record<string, DynamicValue<boolean>>;
  mode: TinyState<Modes>;
  suiteParams: any[];
  testMemoCache: CacheApi<IsolateTest>;
  currentTest?: IsolateTest;
  groupName?: string;
  skipped?: boolean;
  omitted?: boolean;
};

export function useCurrentTest(msg?: string) {
  return SuiteContext.useX(msg).currentTest;
}

export function useGroupName() {
  return SuiteContext.useX().groupName;
}

export function useInclusion() {
  return SuiteContext.useX().inclusion;
}

export function useMode() {
  return SuiteContext.useX().mode();
}

export function useSkipped() {
  return SuiteContext.useX().skipped ?? false;
}

export function useOmitted() {
  return SuiteContext.useX().omitted ?? false;
}

const testMemoCache = cache<IsolateTest>(10);

export function useTestMemoCache() {
  return SuiteContext.useX().testMemoCache;
}

export function useSuiteParams() {
  return SuiteContext.useX().suiteParams;
}
