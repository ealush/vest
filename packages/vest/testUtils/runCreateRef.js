import state from 'state';
import usePending from 'usePending';
import useSuiteId from 'useSuiteId';
import useTestCallbacks from 'useTestCallbacks';
import useTestObjects from 'useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
