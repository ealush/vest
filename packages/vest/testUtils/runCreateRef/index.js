import useTestCallbacks from '../../src/core/produce/useTestCallbacks';
import state from '../../src/core/state';
import useSuiteId from '../../src/core/suite/useSuiteId';
import usePending from '../../src/core/test/lib/pending/usePending';
import useTestObjects from '../../src/core/test/useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
