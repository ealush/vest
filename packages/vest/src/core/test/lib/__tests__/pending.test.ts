import _ from 'lodash';

import expandStateRef from '../../../../../testUtils/expandStateRef';
import runCreateRef from '../../../../../testUtils/runCreateRef';

import VestTest from 'VestTest';
import context from 'ctx';
import { removePending, setPending } from 'pending';
import { usePending, useLagging } from 'stateHooks';


const groupName = 'group_name';

let state, stateRef;

const getStateRef = () => ({ stateRef });

function beforeEachWithContext(cb: () => void) {
  beforeEach(() => context.run({ stateRef }, cb));
}
describe('module: pending', () => {
  let refContent, testObject;

  const runRemovePending = testObject => {
    context.run({ stateRef }, () => removePending(testObject));
  };

  beforeEach(() => {
    stateRef = runCreateRef(state);
    refContent = _.cloneDeep(expandStateRef(stateRef));
    testObject = new VestTest('field_1', jest.fn(), {
      message: 'failure_message',
    });
  });
  describe('export: removePending', () => {
    describe('When testObject it not pending or lagging', () => {
      it('Should keep state unchanged', () => {
        runRemovePending(testObject);
        expect(expandStateRef(stateRef)).toEqual(refContent);
      });
    });

    describe('When testObject is either pending or lagging', () => {
      describe('When in pending', () => {
        beforeEachWithContext(() => {
          const [, setPending] = usePending();

          setPending(pending => {
            return pending.concat(testObject);
          });
        });

        it.withContext(
          'Should remove test from pending',
          () => {
            {
              const [pending] = usePending();
              expect(pending).toContain(testObject);
            }
            runRemovePending(testObject);
            {
              const [pending] = usePending();
              expect(pending).not.toContain(testObject);
            }
            expect(expandStateRef(stateRef)).toMatchSnapshot();
          },
          getStateRef
        );
      });
      describe('When in lagging', () => {
        beforeEachWithContext(() => {
          const [, setLaggin] = useLagging();

          setLaggin(lagging => {
            return lagging.concat(testObject);
          });
        });

        it.withContext(
          'Should remove test from lagging',
          () => {
            {
              const [lagging] = useLagging();
              expect(lagging).toContain(testObject);
            }
            runRemovePending(testObject);
            {
              const [lagging] = useLagging();
              expect(lagging).not.toContain(testObject);
            }
            expect(expandStateRef(stateRef)).toMatchSnapshot();
          },
          getStateRef
        );
      });
    });
  });

  describe('export: setPending', () => {
    let testObjects;

    beforeEach(() => {
      testObjects = Array.from(
        { length: 5 },
        (_, i) =>
          new VestTest(`test_${i}`, jest.fn(), {
            message: 'Some statement string',
          })
      );
      testObjects[0].groupName = groupName;
    });

    it.withContext(
      'Should set supplied test object as pending',
      () => {
        {
          const [pending] = usePending();
          expect(pending).not.toContain(testObjects[0]);
        }
        setPending(testObjects[0]);
        {
          const [pending] = usePending();
          expect(pending).toContain(testObjects[0]);
        }
        expect(expandStateRef(stateRef)).toMatchSnapshot();
      },
      getStateRef
    );

    describe('When a field of the same profile is in lagging array', () => {
      beforeEachWithContext(() => {
        const [, setLagging] = useLagging();

        setLagging(lagging => {
          return lagging.concat(
            testObjects[2], // same fieldName, group = undefined
            testObjects[0], // same fieldName, group = group_name
            testObjects[1] // same fieldName, group = undefined
          );
        });
      });

      it.withContext(
        'Should remove test from lagging array',
        () => {
          {
            const [lagging] = useLagging();
            expect(lagging).toContain(testObjects[0]);
          }
          const added = new VestTest(testObjects[0].fieldName, jest.fn(), {
            groupName: testObjects[0].groupName,
            message: 'failure message',
          });
          setPending(added);
          {
            const [lagging] = useLagging();
            const [pending] = usePending();
            expect(pending).toContain(added);
            expect(lagging).not.toContain(testObjects[0]);
          }
          expect(expandStateRef(stateRef)).toMatchSnapshot();
        },
        getStateRef
      );

      it.withContext(
        'Should add test to pending array',
        () => {
          {
            const [pending] = usePending();
            expect(pending).not.toContain(testObjects[0]);
          }
          setPending(testObjects[0]);
          {
            const [pending] = usePending();
            expect(pending).toContain(testObjects[0]);
          }
        },
        getStateRef
      );

      it.withContext(
        'Should set test as canceled',
        () => {
          expect(testObjects[0].canceled).toBe(false);
          setPending(
            new VestTest(testObjects[0].fieldName, jest.fn(), {
              groupName: testObjects[0].groupName,
              message: 'failure message',
            })
          );
          expect(testObjects[0].canceled).toBe(true);
        },
        getStateRef
      );
    });
  });
});
