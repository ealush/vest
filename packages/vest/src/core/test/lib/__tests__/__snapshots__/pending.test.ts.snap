// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module: pending export: removePending When testObject is either pending or lagging When in lagging Should remove test from lagging 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [],
  "optionalFields": Object {},
  "pending": Array [],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: removePending When testObject is either pending or lagging When in pending Should remove test from pending 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [],
  "optionalFields": Object {},
  "pending": Array [],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: setPending Should set supplied test object as pending 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [],
  "optionalFields": Object {},
  "pending": Array [
    VestTest {
      "canceled": false,
      "failed": false,
      "fieldName": "test_0",
      "groupName": "group_name",
      "id": "4",
      "isWarning": false,
      "message": "Some statement string",
      "testFn": [MockFunction],
    },
  ],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
  },
  "testObjects": Array [],
}
`;

exports[`module: pending export: setPending When a field of the same profile is in lagging array Should remove test from lagging array 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [
    VestTest {
      "canceled": false,
      "failed": false,
      "fieldName": "test_2",
      "id": "12",
      "isWarning": false,
      "message": "Some statement string",
      "testFn": [MockFunction],
    },
    VestTest {
      "canceled": false,
      "failed": false,
      "fieldName": "test_1",
      "id": "11",
      "isWarning": false,
      "message": "Some statement string",
      "testFn": [MockFunction],
    },
  ],
  "optionalFields": Object {},
  "pending": Array [
    VestTest {
      "canceled": false,
      "failed": false,
      "fieldName": "test_0",
      "groupName": "group_name",
      "id": "15",
      "isWarning": false,
      "message": "failure message",
      "testFn": [MockFunction],
    },
  ],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
  },
  "testObjects": Array [],
}
`;
